File: ocor-runtime/src/ocor_runtime/canonical.py
"""RFC 8785 (JCS) JSON canonicalization.

The implementation intentionally does not delegate to ``json.dumps`` for number
or object-key formatting.  JCS uses ECMAScript number serialization and UTF-16
code-unit ordering, both of which differ from Python's JSON defaults at relevant
boundary values.
"""

from __future__ import annotations

import hashlib
import json
import math
from collections.abc import Mapping, Sequence
from typing import Any

from .errors import CanonicalizationError, DigestProviderError

MAX_SAFE_INTEGER = 9_007_199_254_740_991


def _validate_scalar_string(value: str) -> None:
    for character in value:
        codepoint = ord(character)
        if 0xD800 <= codepoint <= 0xDFFF:
            raise CanonicalizationError("lone UTF-16 surrogate is not valid I-JSON")


def _serialize_string(value: str) -> str:
    _validate_scalar_string(value)
    escapes = {
        '"': r'\"',
        "\\": r"\\",
        "\b": r"\b",
        "\t": r"\t",
        "\n": r"\n",
        "\f": r"\f",
        "\r": r"\r",
    }
    encoded: list[str] = ['"']
    for character in value:
        escaped = escapes.get(character)
        if escaped is not None:
            encoded.append(escaped)
        elif ord(character) <= 0x1F:
            encoded.append(f"\\u{ord(character):04x}")
        else:
            encoded.append(character)
    encoded.append('"')
    return "".join(encoded)


def _utf16_sort_key(value: str) -> bytes:
    _validate_scalar_string(value)
    return value.encode("utf-16-be")


def _serialize_float(value: float) -> str:
    if not math.isfinite(value):
        raise CanonicalizationError("NaN and Infinity are not valid I-JSON numbers")
    if value == 0.0:
        return "0"

    sign = "-" if value < 0 else ""
    rendered = repr(abs(value)).lower()
    if "e" in rendered:
        mantissa, exponent_text = rendered.split("e", 1)
        exponent = int(exponent_text)
    else:
        mantissa, exponent = rendered, 0

    if "." in mantissa:
        integer_part, fraction_part = mantissa.split(".", 1)
        digits = integer_part + fraction_part
        decimal_position = len(integer_part) + exponent
    else:
        digits = mantissa
        decimal_position = len(mantissa) + exponent

    while len(digits) > 1 and digits.startswith("0"):
        digits = digits[1:]
        decimal_position -= 1
    while len(digits) > 1 and digits.endswith("0"):
        digits = digits[:-1]

    scientific_exponent = decimal_position - 1
    if scientific_exponent >= 21 or scientific_exponent <= -7:
        coefficient = digits[0]
        if len(digits) > 1:
            coefficient += "." + digits[1:]
        exponent_sign = "+" if scientific_exponent >= 0 else ""
        return f"{sign}{coefficient}e{exponent_sign}{scientific_exponent}"

    if decimal_position <= 0:
        body = "0." + ("0" * -decimal_position) + digits
    elif decimal_position >= len(digits):
        body = digits + ("0" * (decimal_position - len(digits)))
    else:
        body = digits[:decimal_position] + "." + digits[decimal_position:]
    return sign + body


def _serialize_integer(value: int) -> str:
    """Serialize integers only when their value survives the binary64 boundary."""

    if abs(value) <= MAX_SAFE_INTEGER:
        return str(value)
    try:
        binary64 = float(value)
    except OverflowError as exc:
        raise CanonicalizationError(
            "integer is not representable as an IEEE-754 binary64 value"
        ) from exc
    if not math.isfinite(binary64) or int(binary64) != value:
        raise CanonicalizationError(
            "integer is not exactly representable as an IEEE-754 binary64 value"
        )
    return _serialize_float(binary64)


def _serialize(value: Any) -> str:
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, str):
        return _serialize_string(value)
    if isinstance(value, int):
        return _serialize_integer(value)
    if isinstance(value, float):
        return _serialize_float(value)
    if isinstance(value, Mapping):
        if not all(isinstance(key, str) for key in value):
            raise CanonicalizationError("JSON object keys must be strings")
        entries = []
        for key in sorted(value, key=_utf16_sort_key):
            entries.append(f"{_serialize_string(key)}:{_serialize(value[key])}")
        return "{" + ",".join(entries) + "}"
    if isinstance(value, Sequence) and not isinstance(
        value, (str, bytes, bytearray, memoryview)
    ):
        return "[" + ",".join(_serialize(item) for item in value) + "]"
    raise CanonicalizationError(
        f"value of type {type(value).__name__!r} is not an I-JSON value"
    )


def canonicalize_json(value: Any) -> str:
    """Return the RFC 8785 canonical JSON text for a parsed JSON value."""

    return _serialize(value)


def canonicalize(value: Any) -> bytes:
    """Return RFC 8785 canonical JSON encoded as UTF-8 bytes."""

    return canonicalize_json(value).encode("utf-8")


def canonical_sha256(value: Any) -> str:
    """Return a lowercase SHA-256 digest over canonical UTF-8 JSON."""

    try:
        return hashlib.sha256(canonicalize(value)).hexdigest()
    except OSError as exc:
        raise DigestProviderError("SHA-256 provider failure") from exc


def load_i_json(document: str | bytes | bytearray) -> Any:
    """Parse I-JSON while rejecting duplicate keys and non-finite constants."""

    def reject_duplicate_keys(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
        result: dict[str, Any] = {}
        for key, value in pairs:
            if key in result:
                raise CanonicalizationError("duplicate object key")
            result[key] = value
        return result

    def reject_constant(value: str) -> None:
        raise CanonicalizationError(f"non-finite JSON constant: {value}")

    def parse_integer(value: str) -> int | float:
        """Map JSON integers to the I-JSON binary64 domain without host-int drift."""

        try:
            binary64 = float(value)
        except (OverflowError, ValueError) as exc:
            raise CanonicalizationError("invalid I-JSON integer") from exc
        if not math.isfinite(binary64):
            raise CanonicalizationError(
                "integer is not representable as an IEEE-754 binary64 value"
            )
        if abs(binary64) <= MAX_SAFE_INTEGER:
            return int(value)
        return binary64

    try:
        parsed = json.loads(
            document,
            object_pairs_hook=reject_duplicate_keys,
            parse_constant=reject_constant,
            parse_int=parse_integer,
        )
    except CanonicalizationError:
        raise
    except (UnicodeError, ValueError) as exc:
        raise CanonicalizationError("invalid I-JSON document") from exc
    canonicalize(parsed)
    return parsed

File: ocor-runtime/sdk/typescript/src/canonical.ts
/**
 * RFC 8785 (JCS) JSON canonicalization.
 *
 * Ports ocor-runtime/src/ocor_runtime/canonical.py to TypeScript. Unlike the
 * Python port -- which has to hand-replicate ECMAScript's Number::toString
 * algorithm and UTF-16 code-unit string ordering, because Python's own JSON
 * and string defaults differ from both -- this implementation runs on the
 * ECMAScript engine JCS actually specifies, so `String(number)` and the
 * default string comparison operators already ARE the required algorithms.
 * No independent digit-manipulation logic is re-derived here; this is
 * intentional; it is the reason this port is expected to be correct rather
 * than merely tested, and mirrors DEC-166/REM-0007's use of Node.js as the
 * oracle for the Python implementation's own boundary values.
 */

import { createHash } from "node:crypto";
import { CanonicalizationError, DigestProviderError } from "./errors.js";

export const MAX_SAFE_INTEGER = 9_007_199_254_740_991;

export type IJsonValue =
  | null
  | boolean
  | number
  | string
  | readonly IJsonValue[]
  | { readonly [key: string]: IJsonValue };

function hasLoneSurrogate(value: string): boolean {
  for (const codeUnit of value) {
    const codepoint = codeUnit.codePointAt(0);
    if (codepoint !== undefined && codepoint >= 0xd800 && codepoint <= 0xdfff && codeUnit.length === 1) {
      return true;
    }
  }
  return false;
}

function serializeString(value: string): string {
  if (hasLoneSurrogate(value)) {
    throw new CanonicalizationError("lone UTF-16 surrogate is not valid I-JSON");
  }
  let encoded = '"';
  for (const character of value) {
    switch (character) {
      case '"':
        encoded += '\\"';
        break;
      case "\\":
        encoded += "\\\\";
        break;
      case "\b":
        encoded += "\\b";
        break;
      case "\t":
        encoded += "\\t";
        break;
      case "\n":
        encoded += "\\n";
        break;
      case "\f":
        encoded += "\\f";
        break;
      case "\r":
        encoded += "\\r";
        break;
      default: {
        const codepoint = character.codePointAt(0) ?? 0;
        encoded += codepoint <= 0x1f ? `\\u${codepoint.toString(16).padStart(4, "0")}` : character;
      }
    }
  }
  return encoded + '"';
}

function serializeNumber(value: number): string {
  if (!Number.isFinite(value)) {
    throw new CanonicalizationError("NaN and Infinity are not valid I-JSON numbers");
  }
  if (Object.is(value, -0) || value === 0) {
    return "0";
  }
  // Unlike the Python port, there is no separate integer path: JavaScript
  // has one `number` type, already binary64, so an integer outside
  // MAX_SAFE_INTEGER is exactly the same value the finiteness check above
  // already accepted -- no extra representability check is needed.
  // ECMAScript's Number::toString is exactly the JCS-mandated algorithm:
  // shortest round-tripping decimal, scientific notation for
  // exponent >= 21 or <= -7, lowercase "e" with an explicit "+" on the
  // positive exponent. `String(value)` already produces this.
  return String(value);
}

function utf16LessThan(a: string, b: string): boolean {
  // Default `<` on strings already compares by UTF-16 code unit sequence,
  // which is exactly the JCS key-ordering requirement -- no manual encoding
  // to utf-16-be is needed the way the Python port requires.
  return a < b;
}

function isPlainObject(value: unknown): value is { readonly [key: string]: IJsonValue } {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function serialize(value: IJsonValue): string {
  if (value === null) {
    return "null";
  }
  if (value === true) {
    return "true";
  }
  if (value === false) {
    return "false";
  }
  if (typeof value === "string") {
    return serializeString(value);
  }
  if (typeof value === "number") {
    return serializeNumber(value);
  }
  if (Array.isArray(value)) {
    return "[" + value.map((item: IJsonValue) => serialize(item)).join(",") + "]";
  }
  if (isPlainObject(value)) {
    const keys = Object.keys(value);
    for (const key of keys) {
      if (hasLoneSurrogate(key)) {
        throw new CanonicalizationError("lone UTF-16 surrogate is not valid I-JSON");
      }
    }
    keys.sort((a, b) => (utf16LessThan(a, b) ? -1 : utf16LessThan(b, a) ? 1 : 0));
    const entries = keys.map((key) => `${serializeString(key)}:${serialize(value[key] as IJsonValue)}`);
    return "{" + entries.join(",") + "}";
  }
  throw new CanonicalizationError(`value of type ${typeof value} is not an I-JSON value`);
}

/** Return the RFC 8785 canonical JSON text for a parsed JSON value. */
export function canonicalizeJson(value: IJsonValue): string {
  return serialize(value);
}

/** Return RFC 8785 canonical JSON encoded as UTF-8 bytes. */
export function canonicalize(value: IJsonValue): Uint8Array {
  return new TextEncoder().encode(canonicalizeJson(value));
}

/** Return a lowercase SHA-256 digest over canonical UTF-8 JSON. */
export function canonicalSha256(value: IJsonValue): string {
  // canonicalize() runs outside the try so a CanonicalizationError (an
  // invalid input) propagates as itself instead of being mislabeled as a
  // digest-provider failure -- only the hash computation itself is guarded,
  // matching the Python port's `except OSError` (not `except Exception`).
  const bytes = canonicalize(value);
  try {
    return createHash("sha256").update(bytes).digest("hex");
  } catch {
    throw new DigestProviderError("SHA-256 provider failure");
  }
}
